test/concurrency/lock_manager_test.cpp

```diff
@@ -152,119 +152,51 @@ void UpgradeTest() {
-  std::sort(edges.begin(), edges.end());
+  auto wait_die_task = [&]() {
+    // younger transaction acquires lock first
+    Transaction txn_die(id_die);
+    txn_mgr.Begin(&txn_die);
+    bool res = lock_mgr.LockExclusive(&txn_die, rid);
+    EXPECT_TRUE(res);
 
-  for (int i = 0; i < num_edges; i++) {
-    EXPECT_EQ(edges[i], lock_mgr_edges[i]);
-  }
-}
+    CheckGrowing(&txn_die);
+    CheckTxnLockSize(&txn_die, 0, 1);
 
-TEST(LockManagerTest, DISABLED_BasicCycleTest) {
-  LockManager lock_mgr{}; /* Use Deadlock detection */
-  TransactionManager txn_mgr{&lock_mgr};
+    // wait for txn 0 to call lock_exclusive(), which should wound us
+    std::this_thread::sleep_for(std::chrono::milliseconds(300));
 
-  /*** Create 0->1->0 cycle ***/
-  lock_mgr.AddEdge(0, 1);
-  lock_mgr.AddEdge(1, 0);
-  EXPECT_EQ(2, lock_mgr.GetEdgeList().size());
+    CheckAborted(&txn_die);
 
-  txn_id_t txn;
-  EXPECT_EQ(true, lock_mgr.HasCycle(&txn));
-  EXPECT_EQ(1, txn);
+    // unlock
+    txn_mgr.Abort(&txn_die);
+  };
 
-  lock_mgr.RemoveEdge(1, 0);
-  EXPECT_EQ(false, lock_mgr.HasCycle(&txn));
-}
+  Transaction txn_hold(id_hold);
+  txn_mgr.Begin(&txn_hold);
 
-TEST(LockManagerTest, DISABLED_BasicDeadlockDetectionTest) {
-  LockManager lock_mgr{};
-  cycle_detection_interval = std::chrono::milliseconds(500);
-  TransactionManager txn_mgr{&lock_mgr};
-  RID rid0{0, 0};
-  RID rid1{1, 1};
-  auto *txn0 = txn_mgr.Begin();
-  auto *txn1 = txn_mgr.Begin();
-  EXPECT_EQ(0, txn0->GetTransactionId());
-  EXPECT_EQ(1, txn1->GetTransactionId());
-
-  std::thread t0([&] {
-    // Lock and sleep
-    bool res = lock_mgr.LockExclusive(txn0, rid0);
-    EXPECT_EQ(true, res);
-    EXPECT_EQ(TransactionState::GROWING, txn0->GetState());
-    std::this_thread::sleep_for(std::chrono::milliseconds(100));
-
-    // This will block
-    lock_mgr.LockExclusive(txn0, rid1);
-
-    lock_mgr.Unlock(txn0, rid0);
-    lock_mgr.Unlock(txn0, rid1);
-
-    txn_mgr.Commit(txn0);
-    EXPECT_EQ(TransactionState::COMMITTED, txn0->GetState());
-  });
-
-  std::thread t1([&] {
-    // Sleep so T0 can take necessary locks
-    std::this_thread::sleep_for(std::chrono::milliseconds(50));
-    bool res = lock_mgr.LockExclusive(txn1, rid1);
-    EXPECT_EQ(res, true);
-    EXPECT_EQ(TransactionState::GROWING, txn1->GetState());
-
-    // This will block
-    try {
-      res = lock_mgr.LockExclusive(txn1, rid0);
-      EXPECT_EQ(TransactionState::ABORTED, txn1->GetState());
-      txn_mgr.Abort(txn1);
-    } catch (TransactionAbortException &e) {
-      // std::cout << e.GetInfo() << std::endl;
-      EXPECT_EQ(TransactionState::ABORTED, txn1->GetState());
-      txn_mgr.Abort(txn1);
-    }
-  });
+  // launch the waiter thread
+  std::thread wait_thread{wait_die_task};
 
-  // Sleep for enough time to break cycle
-  std::this_thread::sleep_for(cycle_detection_interval * 2);
+  // wait for txn1 to lock
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
This is a little nitpicky because I'm sure this solution works in practice, but what do you think about using a latch primitive for this instead? The standard library doesn't include a latch until C++20, but we could roll our own with a `std::atomic` very easily, and it removes any potential for nondeterminism.